fdbserver/storageserver.actor.cpp

```diff
@@ -218,100 +218,93 @@ struct MoveInUpdates {
-		for (const auto& [version, vur] : restored) {
-			ASSERT(version == vur.version);
-			if (version > begin) {
-				self->updates.push_back(vur);
-			}
-		}
-
-		while (!tmp.empty()) {
-			if (self->updates.empty() || tmp.front().version > self->updates.back().version) {
-				TraceEvent(SevDebug, "MoveInUpdatesRestoreAddingNewUpdates", self->id)
-				    .detail("Version", tmp.front().version)
-				    .detail("Size", tmp.front().mutations.size());
-				self->updates.push_back(tmp.front());
-			}
-			tmp.pop_front();
-		}
-
-		return Void();
-	}
+	ACTOR static Future<Void> loadUpdates(MoveInUpdates* self, Version begin, Version end);
 
 	Key getPersistKey(const Version version, const int idx) const;
 };
 
-bool MoveInUpdates::hasNext() const {
-	return !updates.empty();
-}
+ACTOR Future<Void> MoveInUpdates::loadUpdates(MoveInUpdates* self, Version begin, Version end) {
+	if (begin >= end) {
+		self->spilled = MoveInUpdatesSpilled::False;
+		return Void();
+	}
 
-std::vector<Standalone<VerUpdateRef>> MoveInUpdates::next(const int byteLimit) {
-	std::vector<Standalone<VerUpdateRef>> res;
-	int size = 0;
-	if (!this->fail) {
-		while (!updates.empty()) {
-			if (updates.front().version > this->lastAppliedVersion) {
-				res.push_back(updates.front());
-				size += res.back().mutations.expectedSize();
-			}
-			updates.pop_front();
-			if (size > byteLimit) {
-				break;
+	const Key beginKey = persistUpdatesKey(self->id, begin), endKey = persistUpdatesKey(self->id, end);
+	TraceEvent(self->logSev, "MoveInUpdatesLoadBegin", self->id)
+	    .detail("BeginVersion", begin)
+	    .detail("EndVersion", end)
+	    .detail("BeginKey", beginKey)
+	    .detail("EndKey", endKey);
+	ASSERT(beginKey < endKey);
+	RangeResult res = wait(self->store->readRange(KeyRangeRef(beginKey, endKey),
+	                                              SERVER_KNOBS->FETCH_SHARD_UPDATES_BYTE_LIMIT,
+	                                              SERVER_KNOBS->FETCH_SHARD_UPDATES_BYTE_LIMIT));
+	std::vector<Standalone<VerUpdateRef>> restored;
+	for (int i = 0; i < res.size(); ++i) {
+		BinaryReader rd(res[i].key.removePrefix(self->range.begin), Unversioned());
+		uint64_t uv;
+		rd >> uv;
+		const Version version = static_cast<Version>(fromBigEndian64(uv));
+		Standalone<VerUpdateRef> vur =
+		    BinaryReader::fromStringRef<Standalone<VerUpdateRef>>(res[i].value, IncludeVersion());
+		ASSERT(version == vur.version);
+		TraceEvent(self->logSev, "MoveInUpdatesLoadedMutations", self->id)
+		    .detail("Version", version)
+		    .detail("Mutations", vur.mutations.size());
+		restored.push_back(std::move(vur));
+	}
+
+	if (!res.more) {
+		for (int i = restored.size() - 1; i >= 0; --i) {
+			if (self->updates.empty() || restored[i].version < self->updates.front().version) {
+				self->updates.push_front(std::move(restored[i]));
 			}
 		}
+		self->spilled = MoveInUpdatesSpilled::False;
+	} else {
+		ASSERT(self->spillBuffer.empty());
+		std::swap(self->spillBuffer, restored);
```

Comment:
Set `self->spilled` to true here?

Reply:
Yes, setting it true here is correct, but not actually necessary since `loadUpdates` is called only if `self->spilled` is true. I added an ASSERT at the beginning of the function.